pulsar-client/src/main/java/org/apache/pulsar/client/impl/BatchMessageContainerImpl.java

```diff
@@ -137,6 +137,9 @@ public boolean add(MessageImpl<?> msg, SendCallback callback) {
         if (producer != null) {
             ProducerImpl.LAST_SEQ_ID_PUSHED_UPDATER.getAndUpdate(producer, prev -> Math.max(prev, msg.getSequenceId()));
         }
+        if (!messageMetadata.hasSchemaId() && msg.getSchemaId() != null) {
+            messageMetadata.setSchemaId(msg.getSchemaId());
+        }
```

Comment:
It should be moved to the block of 

```java
        if (++numMessagesInBatch == 1) {
```

because we only need to set the schema id for the 1st message in the batch.

The schema id validation between messages in the same batch has already been performed in `hasSameSchema`, which is called in `canAddToCurrentBatch`

Reply:
+1